[limkongkiat] iP

limkongkiat opened the pull request.

limkongkiat committed:
Change bot name to Tyrone

limkongkiat committed:
Add echo function

limkongkiat committed:
Implement Add Task and List Task features

Tasks are added, stored and listed using a new class
TaskList in TaskList.java for greater neatness of code

limkongkiat committed:
Add mark/unmark task feature

Create new class Task in Task.java to record
description and whether task is done

limkongkiat committed:
Make minor tweaks to comply with coding standard

limkongkiat committed:
Add CommandHandler class to handle commands

Extracting code to parse and handle commands from main to a new class
allows for greater readability of main code.

Let's extract code to parse and handle commands to a new class CommandHandler.

limkongkiat committed:
Add Todo class to keep track of Todo tasks

limkongkiat committed:
Add Deadline class to track deadlines

limkongkiat committed:
Add Event class to track event tasks

limkongkiat committed:
Refactor handleCommand by extracting multiple methods

Before this, handleCommand was a very long method with poor readability.

Let's extract out code in handleCommand into multiple methods, with one method
to handle each command type. This shortens the handleCommand method which
makes it much more readable.

limkongkiat committed:
Add input.txt to allow for semi-automated testing

limkongkiat committed:
Amend Tyrone main method to improve readability.

Main path (handleCommand) was initially indented in an if block
which checks if the bye command has been entered.

Let's amend the code to reduce the level of indentation of the main path.

TVageesan commented on src/main/java/CommandHandler.java:
```diff
@@ -0,0 +1,81 @@
+        return command.equals("bye");
+    }
+
+    public static void handleCommand(String command) {
+        String[] dissectedCommand = command.split(" ");
+        if (command.equals("list")) {
+            handleList();
+        } else if (dissectedCommand[0].equals("mark")) {
+            handleMark(dissectedCommand);
+        } else if (dissectedCommand[0].equals("unmark")) {
+            handleUnmark(dissectedCommand);
+        } else if (dissectedCommand[0].equals("todo")) {
+            handleTodo(command);
+        } else if (dissectedCommand[0].equals("deadline")) {
+            handleDeadline(command);
+        } else if (dissectedCommand[0].equals("event")) {
+            handleEvent(command);
+        } else {
+            handleTask(command);
+        }
+    }
```
This could be done in a more readable way by using a switch case expression.

TVageesan replied:
Could this be expressed as a switch case instead?

TVageesan commented on src/main/java/CommandHandler.java:
```diff
@@ -0,0 +1,81 @@
+        }
+    }
+
+    private static void handleList() {
```
Is handleList too vague of a function name to describe its purpose?

TVageesan commented on src/main/java/CommandHandler.java:
```diff
@@ -0,0 +1,81 @@
+    }
+
+    private static void handleMark(String[] dissectedCommand) {
+        int taskId = Integer.parseInt(dissectedCommand[1]) - 1;
```
Could this line be abstracted into a function such as parseIndex to make it more readable?

TVageesan commented on src/main/java/CommandHandler.java:
```diff
@@ -0,0 +1,81 @@
+    }
+
+    private static void handleEvent(String command) {
+        String description = command.substring(command.indexOf(" ") + 1, command.indexOf(" /from"));
```
Could this be abstracted to a helper function like extractArguments?

TVageesan commented on src/main/java/CommandHandler.java:
```diff
@@ -0,0 +1,81 @@
+public class CommandHandler {
```
CommandHandler is named well, making the class's purpose clear and also following PascalCase convention.

limkongkiat committed:
Add error handling for Deadline and Event tasks

Program would crash when Deadline/Event tasks were input
improperly (i.e. task deadline/start/end was missing or input without
proper flag)

Let's add try-catch statements to handle the exceptions arising from
these situations, and print out the proper format for inputting Deadline/Event
tasks when user inputs Deadline/Event tasks with the incorrect format.

limkongkiat committed:
Change handleTask method to handleUnknown to handle unrecognized commands

limkongkiat committed:
Add error handling for mark/unmark tasks

Code crashes when mark/unmark commands were input with no index or
non-numeric index.

Let's add try-catch statements to handle missing index or non-numeric index
being input into the program, which will also print an error message informing the user
of the mistake they made in inputting the command.

limkongkiat committed:
Add error handling to handle empty input fields for tasks

Program allows users to successfully log tasks with empty descriptions,
deadlines, start times or end times.

This is bad as it can potentially allow for tasks with empty fields to
be logged which may confuse the user later on.

Let's create a EmptyFieldException class to flag empty deadline/
start date/end date inputs. handleTodo, handleDeadline and handleEvent
methods will throw a EmptyFieldException upon detecting empty field
inputs, which triggers the program to print an error message informing
the user not to input empty fields.

A new EmptyFieldException class was used to clearly signal to future
code readers that the code is meant to handle errors regarding empty
fields.

limkongkiat committed:
Refactor hasNonWhitespaceChar method to hasOnlyWhitespaceChar method

Program checks for empty input fields (description/deadline/start/end)
by checking if hasNonWhitespaceChar returns false for these fields.

This creates unnecessary confusion for future readers of the code.

Let's refactor the hasNonWhitespaceChar method to hasOnlyWhitespaceChar
such that the program detects empty fields when hasOnlyWhitespaceChar
returns true.

This helps make reading the code less confusing by removing
negation signs in the conditional statements when checking for
empty fields.

limkongkiat committed:
Merge branch 'branch-Level-5'

limkongkiat committed:
Sort files into different packages for easier organization

limkongkiat committed:
Merge branch 'branch-A-Packages'

limkongkiat committed:
Change TaskList to store tasks using ArrayList instead of array

TaskList uses a fixed-sized array to store tasks.

This limits the number of tasks we can store in our list and is
inefficient in terms of memory usage as we cannot dynamically size
the array.

Let's replace the array of tasks with an ArrayList to store the tasks.

This lets us store and unlimited number of tasks and allows more
efficient use of memory through dynamic sizing of the ArrayList.

limkongkiat committed:
Implement Delete Task feature

limkongkiat committed:
Add getAllTaskDetails method to TaskList class

Program has no means of retrieving all tasks logged in TaskList,
making writing the list of tasks to a new file difficult.

Let's create a method getAllTaskDetails that compiles the details of
all logged tasks and returns it as a single string.

This allows the program to write all tasks onto an external file
easily.

limkongkiat committed:
Add methods to create and update save file

Program does not save tasks logged into it, meaning all data is lost
once the program is closed.

Let's create a FileReadWriter class with methods to create a new file
to store the logged tasks and write the list of all logged tasks to it
whenever the list of tasks is updated.

Feature of reading from the save file is a work in progress.

limkongkiat committed:
Implement feature to save tasks in external .txt file

limkongkiat committed:
Implement Read Task list from save file feature

Program saves list of tasks to external file but does not read from
the file to access the saved task list.

Let's add an initTaskListFromSaveFile method in FileReadWriter class
which populates TaskList according to the list of tasks logged in
the save file.

This allows the program to initialize its list of tasks according to
the list of tasks as of the program's last closure. FileReadWriter has
also been added to a new savemanager package for better organization of
files.

limkongkiat committed:
Update .gitignore to ignore save file

limkongkiat committed:
Merge branch 'branch-Level-6'

limkongkiat committed:
Merge branch 'branch-Level-7'

limkongkiat committed:
Fix getAllTaskDetails method to work with ArrayList task list

limkongkiat committed:
Create jar manifest

thomasjlalba commented on src/main/java/tyrone/command/CommandHandler.java:
```diff
@@ -0,0 +1,157 @@
+
+    private static void handleEvent(String command) {
+        try {
+            String description = command.substring(command.indexOf(" ") + 1, command.indexOf(" /from"));
```
Maybe you could avoid magic numbers here along with a few other areas in the code!

thomasjlalba commented on src/main/java/tyrone/command/CommandHandler.java:
```diff
@@ -0,0 +1,157 @@
+        }
+    }
+
+    private static boolean hasOnlyWhitespaceChar(String string) {
```
Maybe you can take a look at string.isBlank()

thomasjlalba commented on src/main/java/tyrone/savemanager/FileReadWriter.java:
```diff
@@ -0,0 +1,100 @@
+        case 'E':
+            parseEvent(line);
+            break;
+        }
```
Maybe including a default case for error handling would be good.

thomasjlalba commented on src/main/java/tyrone/Tyrone.java:
```diff
@@ -0,0 +1,26 @@
+        System.out.println("Hello! I'm Tyrone.\nWhat can I do for you?");
+
+        while (true) {
+            String command = scanner.nextLine();
+            if(CommandHandler.isExitCommand(command)) {
+                break;
+            }
+            CommandHandler.handleCommand(command);
```
Good job on extracting it to smaller methods

thomasjlalba commented on src/main/java/tyrone/command/CommandHandler.java:
```diff
@@ -0,0 +1,157 @@
+        String[] dissectedCommand = command.split(" ");
+        if (command.equals("list")) {
+            handleList();
+        } else if (dissectedCommand[0].equals("mark")) {
```
Maybe you could refactor `dissectedCommand[0]` to indicate it the command, and the rest are arguments?

thomasjlalba reviewed (commented):
Other than some refactoring that can make it better, great work on following most coding standards!

limkongkiat committed:
Refactor CommandHandler to InputHandler

Entire user input is saved as command variable, which is handled by
CommandHandler class.

Let's refactor our code to relabel the user's input as input, which
allows us to name the command type of the user input (todo, deadline,
etc.) as 'command'.

limkongkiat committed:
Refactor magic literals to constant variables

limkongkiat committed:
Replace hasOnlyWhitespaceChar method with isBlank method

limkongkiat committed:
Change format of reading/writing data in Tyrone.txt

Data format of reading/writing Tyrone.txt allows for errors to arise
from the way the user writes the task descriptions (e.g. if description
contains "(by: " substring.)

Let's change the format in which data is written to and read from
Tyrone.txt to greater resemble the user input format so as to
reduce likelihood of an erronous update of TaskList from Tyrone.txt
arising from format of user-input description.

limkongkiat committed:
Update print messages of methods to parse Tyrone.txt

limkongkiat committed:
Refactor printList method to return String instead of void

The printList method prints the list of tasks directly.

Let's :
* refactor the method to return a String of the list of tasks
instead. This allows us to use a Ui class to print the list of tasks
instead for better adherence to the OOP Single Responsibility
Principle.
* rename method to getAllTaskDetails to better describe newly
refactored method's function
* rename original getAllTaskDetails method to getAllTaskSaveRecords to
prevent the 2 methods from sharing the same name

limkongkiat committed:
Add Ui class to handle printing to UI and receiving user inputs

limkongkiat committed:
Remove redundant import in Ui class

limkongkiat committed:
Add Find task feature

limkongkiat committed:
Refactor indentation as constant literal

In handleMark, handleUnmark and handleDelete methods, the task that
was marked/unmarked/deleted is printed with a 2-space indent.

Let's refactor this 2-space indent as a constant literal to better
adhere to Java coding standards.

limkongkiat committed:
Update handleFind method to handle missing keyword case

limkongkiat committed:
Add JavaDoc comments for classes and methods

limkongkiat committed:
Merge pull request #1 from limkongkiat/branch-Level-9

Add Find feature (Level-9)

limkongkiat committed:
Merge pull request #2 from limkongkiat/branch-A-JavaDoc

Add JavaDoc comments for classes and methods

limkongkiat committed:
Add JavaDoc comments for handleFind method

limkongkiat committed:
Rename command variable to input for some methods in Parser

Storing user input in variable named command could be confusing
with command in handleInput method, which refers to the first word
of the user input (list, event, etc.).

Let's rename the command variable in methods handleDeadline,
handleEvent and handleTodo to input to reduce likelihood of readers
of code getting confused.

limkongkiat committed:
Modify classes to contain and use instances of each other

Chatbot uses static methods from TaskList, Ui, Storage and Parser for
its various functions. This is not very good adherence to OOP.

Let's:
* Convert most methods in Tyrone/Ui/Parser/Storage/TaskList from
static to normal methods.
* Create instances of Ui, Storage, Parser and TaskList within Tyrone,
such that the Tyrone class calls methods from these instances of the
classes.

This is a stronger adherence to OOP.

limkongkiat committed:
Use Ui class to print error messages in Storage class methods

limkongkiat committed:
Add getNumOfTasks method to TaskList

Let's add a method to TaskList class to return the number of tasks
currently in the task list. This allows the chatbot to notify the user
how many tasks are currently logged.

limkongkiat committed:
Update methods for adding/deleting tasks to print number of tasks

Let's modify the methods used to add/delete tasks to print the number
of tasks current in the task list after adding/deleting the specified
task.

limkongkiat committed:
Update JavaDoc comments for Parser constructor

limkongkiat committed:
Rename START_INDEX_OFFSET_DESC for greater clarity

Let's rename START_INDEX_OFFSET_DESC, a constant variable to store
the number of characters after the first ' ' character in the user
input which the task description begins, to
START_INDEX_OFFSET_DESCRIPTION. This is for the sake of reduced
ambiguity.

limkongkiat committed:
Refactor Ui class into new ui package

limkongkiat committed:
List imported classes explicitly in Storage class

limkongkiat committed:
Add User Guide as README

limkongkiat committed:
Update User Guide to ensure proper rendering

Table at bottom of README is not rendering properly on the Github page.

Let's add a blank line before our table to ensure that it renders
properly on the Github page.

limkongkiat committed:
Add JAR file release link to User Guide

limkongkiat committed:
Make minor style changes to User Guide

limkongkiat committed:
Extract magic literals in Parser and Storage as constants

Flags to denote task deadlines/start/event are written as magic
literals in Parser and Storage functions. This violates Java coding
standards and could potentially confuse readers of the code.

Let's:
* extract out these magic literals as constants
* modify the methods in these classes such that constants checked for
are more consistent throughout the code, allowing for greater
reusability of the newly created constants

Extracting these magic literals into constants improves clarity of
the code to readers.

limkongkiat committed:
Extract shared constants to new Interface class TaskInputConstants

Parser and Storage classes share several common constants which
causes code duplication.

Let's extract these constants to a new Interface class
TaskInputConstants, which is then implemented by Parser and Storage.

An Interface class was used instead of a regular public class so that
only Parser and Storage can access these constants.

limkongkiat committed:
Make minor improvement to code formatting

limkongkiat committed:
Add remark in User Guide that Find feature is case-sensitive

limkongkiat committed:
Add contents page to User Guide

limkongkiat committed:
Make minor edits to code style
